rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,499 @@
+
+- [downlevel-dts] allows you to take a `.d.ts` file which is not valid for an earlier version of TypeScript (e.g. the changes to class field emit mentioned in [<b>Breaking Changes to Type Definitions</b>](#breaking-changes-to-type-definitions)), and emit a version which *is* compatible with that version. Specifically: it currently generates types compatible with TypeScript 3.4.[[see note 2](#notes)]
+- [typesVersions] allow you to specify a specific set of type definitions (which may consist of one or more `.d.ts` files) which correspond to a specific TypeScript version
+
+[downlevel-dts]: https://github.com/sandersn/downlevel-dts
+[typesVersions]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#version-selection-with-typesversions
+
+The recommended flow will be as follows:[[see note 2](#notes)]
+
+1.  Add `downlevel-dts`, `npm-run-all`, and `rimraf`  to your dev dependencies:
+
+    ```sh
+    npm install --save-dev downlevel-dts npm-run-all rimraf
+    ```
+    
+    or 
+    
+    ```sh
+    yarn add --dev downlevel-dts npm-run-all rimraf
+    ```
+
+2.  Update the `scripts` key in `package.json`  to generate downleveled types generated by running `downlevel-dts` on the output of `ember-cli-typescript`’s  `ts:precompile` command, and to clean up the results after publication:
+
+    ```diff
+    {
+      "scripts": {
+    -   "prepublishOnly": "ember ts:precompile",
+    +   "prepublish:types": "ember ts:precompile",
+    +   "prepublish:downlevel": "downlevel-dts . ts3.4",
+    +   "prepublishOnly": "run-s prepublish:types prepublish:downlevel",
+    -   "postpublish": "ember ts:clean",
+    +   "clean:ts": "ember ts:clean",
+    +   "clean:downlevel": "rimraf ./ts3.4",
+    +   "clean": "npm-run-all --aggregate-output --parallel clean:*",
+    +   "postpublish": "npm run clean",
+      }
+    }
+    ```
+
+3.  Add a `typesVersions` key to `package.json`, with the following contents, where instead of `3.9` the addon should always supply the current maximum supported TypeScript version:
+
+    ```json
+    {
+      "types": "index.d.ts",
+      "typesVersions": {
+        "<3.9": { "*": "ts3.4/*" }
+      }
+    }
+    ```
+
+    This will tell TypeScript how to use the types generated during `ember ts:precompile`. Note that we explicitly include the `types` key so TypeScript will fall back to the defaults for 3.9 and higher.
+
+4.  If using the `files` key in `package.json` to specify files to include (unusual but not impossible for TypeScript-authored addons), add `ts3.4` to the list of entries.
+
+Now consumers using older versions of TypeScript will be buffered from the breaking changes in type definition emit.
+
+These changes to `package.json` may be able to be automated to some degree; Typed Ember should explore a blueprint for updating the supported TypeScript version.
```

Comment:
After doing the work on ember-modifier, and supporting it for ember-concurrency, I'm strongly inclined to agree. I'd be inclined to explore it in an independent package, since those three pieces really have no logical connection to ember-cli-typescript as such. I'll note that we intend to build that tooling here.

Reply:
Replaced this with a parenthetical note on tooling here in a1784c6.